### PR TITLE
Add support for PreAuthKey tags/automatic tagging

### PR DESCRIPTION
**Pull request by tsujamin**
- [X] read the [CONTRIBUTING guidelines](README.md#contributing)
- [x] raised a GitHub issue or discussed it on the projects chat beforehand
- [X] added unit tests
- [X] added integration tests
- [ ] updated documentation if needed
- [x] updated CHANGELOG.md

These changes add the ability to automatically tag nodes when activated with a preauthkey as is available in SaaS

CreatePreAuthKey and like APIs were expanded to optionally accept a list of tags and the CLI arguments were updated. The database schema was also updated with a new table to store tags associated with preauth keys.

Keys associated to a preauthkey are applied to a new or existing Machine's ForcedKeys when registered/refreshed

### CODE REVIEW

**Comment by tsujamin** [comment]
just to note that there's a small test break I'll need to push a change to depending on which order this and #763 get merged (if either is accepted

**Comment by tsujamin** [comment]
fixed broken test from merge

**Review by restanrm — COMMENTED** [review]
could you add a unit test to check that a node registered with such PreAuthKey have the proper Tags associated ?

**Comment by tsujamin** [comment]
np will have a crack tomorrow - just noticed my local linting is broken too so will address both

**Comment by restanrm** [comment]
> np will have a crack tomorrow - just noticed my local linting is broken too so will address both

Thanks ! Maybe you'll want to wait for an owner review first :smile:

**Comment by restanrm** [comment]
> > np will have a crack tomorrow - just noticed my local linting is broken too so will address both
> 
> Thanks ! Maybe you'll want to wait for an owner review first smile

Sorry this comment was for another PR. No need for approval to add tests they are always welcome :smile:

**Comment by tsujamin** [comment]
haha all good, fwiw just about to push a fix for the linting issues but not the unit test as the point where tags is applied is in `protocol_common.go` as part of machine registration/refresh as this most closely follows SaaS behaviour.

Let me know if an equivalent integration test is desired - just a bit of extra effort to stand up a new test case in there 🥲

File: protocol_common.go [hunk 1]
```diff
@@ -355,6 +356,25 @@ func (h *Headscale) handleAuthKeyCommon(
 
 			return
 		}
+
+		aclTags := pak.toProto().AclTags
+		if len(aclTags) > 0 {
+			// This conditional preserves the existing behaviour, although SaaS would reset the tags on auth-key login
+			err = h.SetTags(machine, aclTags)
```
Comment (kradalby) [inline protocol_common.go:360]: If this _overrides_ the tags, then technically if the list is empty, we should still set it? as in, we should set it to no tags?

File: protocol_common.go [hunk 2]
```diff
@@ -355,6 +356,25 @@ func (h *Headscale) handleAuthKeyCommon(
...
+		}
+
+		if err != nil {
```
Comment (kradalby) [inline protocol_common.go:363]: This error check should be inside the `if len(aclTags)` closure so we don't risk any weird scoping behaviour.

File: cmd/headscale/cli/preauthkeys.go [hunk 3]
```diff
@@ -96,6 +108,15 @@ var listPreAuthKeys = &cobra.Command{
...
 			}
 
+			var aclTags string
+
+			if len(key.AclTags) > 0 {
+				for _, tag := range key.AclTags {
+					aclTags += "," + tag
+				}
+				aclTags = strings.TrimLeft(aclTags, ",")
+			}
```
Comment (kradalby) [inline cmd/headscale/cli/preauthkeys.go:118]: This `len` check is strictly not necessary, this should work:
```suggestion
			aclTags := ""

			for _, tag := range key.AclTags {
				aclTags += "," + tag
			}
			aclTags = strings.TrimLeft(aclTags, ",")
```

Comment (kradalby) [reply cmd/headscale/cli/preauthkeys.go:118]: Since the loop will not run if the list has 0 items, its safe to not check it.

File: grpcv1.go [hunk 4]
```diff
@@ -106,11 +106,24 @@ func (api headscaleV1APIServer) CreatePreAuthKey(
...
+		for _, tag := range request.AclTags {
+			err := validateTag(tag)
+
+			if err != nil {
+				return &v1.CreatePreAuthKeyResponse{
+					PreAuthKey: nil,
+				}, status.Error(codes.InvalidArgument, err.Error())
+			}
+		}
+	}
```
Comment (kradalby) [inline grpcv1.go:119]:
```suggestion
	for _, tag := range request.AclTags {
		err := validateTag(tag)

		if err != nil {
			return &v1.CreatePreAuthKeyResponse{
				PreAuthKey: nil,
			}, status.Error(codes.InvalidArgument, err.Error())
		}
	}
```
We don't need to check for empty lists as `range` will just not have anything to iterate over.

File: preauth_keys.go [hunk 5]
```diff
@@ -186,5 +234,11 @@ func (key *PreAuthKey) toProto() *v1.PreAuthKey {
 		protoKey.CreatedAt = timestamppb.New(*key.CreatedAt)
 	}
 
+	if len(key.ACLTags) > 0 {
+		for idx := range key.ACLTags {
+			protoKey.AclTags[idx] = key.ACLTags[idx].Tag
+		}
+	}
```
Comment (kradalby) [inline preauth_keys.go:241]:
```suggestion
	for idx := range key.ACLTags {
		protoKey.AclTags[idx] = key.ACLTags[idx].Tag
	}
```

**Review by kradalby — COMMENTED** [review]
Good work!, a few nits, one questions, otherwise good.

File: protocol_common.go (hunk 1 above)
Comment (tsujamin) [reply protocol_common.go:360]: So this handles some divergent behaviour between SaaS and headscale. I believe it's possible in headscale for a machine to re-auth with a pre-auth key and keep its previously set tags, whereas upstream would reset the tags in my testing.

This logic tries to keep the best of both worlds, when a machine is re-authenticated with a pak:
  * if the pak has no associated tags, the machine keeps its previous tags (the current, non SaaS behaviour)
  * if the pak has tags, these tags replace any previously set on the machine (which would more closely align with a user's *intent*)

Removing the conditional would mean that, when a machine is re-authenticated with a pak, it would lose any previous tags. This is what SaaS does but I didn't want to force a change in headscales behaviour

tl;dr I feel this is messy but its somewhat behaviour preserving

File: cmd/headscale/cli/preauthkeys.go (hunk 3 above)
Comment (tsujamin) [reply cmd/headscale/cli/preauthkeys.go:118]: :+1: ty this was me not understanding go at this point of the journey haha

File: protocol_common.go (hunk 1 above)
Comment (kradalby) [reply protocol_common.go:360]: ok, make sense, we can live with that for now, and if we decide to change it, we will have to mark it as breaking.

**Comment by tsujamin** [comment]
@kradalby thanks for the review - have merged in your recommendations

**Comment by kradalby** [comment]
Seems like there is a lint error, and could you add changelog entry for this one as well?

**Review by kradalby — APPROVED** [review]
